templates/fine-tune-llm_v2/cookbooks/lora_vs_full_parameter/README.md

```diff
@@ -0,0 +1,98 @@
+
+### How to launch a LoRA fine-tuning job
+
+
+```python
+# [Optional] You can set the WandB API key to track model performance
+# !export WANDB_API_KEY={YOUR_WANDB_API_KEY}
+
+# Fine-tune Llama 3 8B with LoRA
+!python main.py llama-3-8b-lora.yaml
+```
+
+### How to launch a full-parameter fine-tuning job
+
+We advise against running this end-to-end, because it will take many hours on a default cluster.
+If you want to run this full-parameter fine-tuning end-to-end, consider configuring your workspace to use more capable GPUs like A100s or better.
+
+
+```python
+# [Optional] You can set the WandB API key to track model performance
+# !export WANDB_API_KEY={YOUR_WANDB_API_KEY}
+
+# Fine-tune Llama 3 8B with full-parameter fine-tuning
+!python main.py llama-3-8b-full-parameter.yaml
+```
+
+## Comparison of configurable parameters
+
+### LoRA fine-tuninng
+
+In [our blogpost](https://www.anyscale.com/blog/fine-tuning-llms-lora-or-full-parameter-an-in-depth-analysis-with-llama-2), you'll find information on what the parameters mean. Here is a snippet from the yaml used above with explanatory comments for the most interesting parameters.
+
+```yaml
+lora_config:
+  # Determines the rank of the matrices that we fine-tune. Higher rank means more parameters to fine-tune. Increasing the rank gives you diminishing returns.
+  r: 8
+  # Scales the learnt LoRA weights. A value 16 is common practice and is not advised to be fine-tuned.
+  lora_alpha: 16
+  # Rate at which LoRA weights are dropped out. Can act as a regularizer.
+  lora_dropout: 0.05
+  # The modules of the LLM that we want to fine-tune with LoRA. 
+  target_modules:
+    - q_proj
+    - v_proj
+    - k_proj
+    - o_proj
+    - gate_proj
+    - up_proj
+    - down_proj
+    - embed_tokens
+    - lm_head
+```
+
+You generally don't need to change any of these parameters. In our experience, there is little or nothing to be gained from that.
+We advise to fine-tune LoRA with a learning rate of about 1e-4. You can increase it slightly if training is stable enough.
+
+### Full-parameter fine-tuning
+
+Full-parameter fine-tuning requires the same config as LoRA fine-tuning, but the `lora_config` part should be omited.
+We advise to use a learning rate of about 1e-5 here. You can increase it slightly if training is stable enough.
+
+## FAQ
+
+### Should I use LoRA or full-parameter fine-tuning?
+
+There is no general answer to this but here are some things to consider:
+
+- The quality of the fine-tuned models will, in most cases, be comparable if not the same
+- LoRA shines if...
+    - ... you want to serve many fine-tuned models at once yourself
+    - ... you want to rapidly experiment (because fine-tuning, downloading and serving the model take less time)
+- Full-parameter shines if...
+    - ... you want to make sure that your fine-tuned model has the maximum quality
+    - ... you want to serve only one fine-tuned version of the model
```

Comment:
```suggestion
- LoRA shines if:
    - You want to serve many fine-tuned models at once yourself
    - You want to rapidly experiment (because fine-tuning, downloading and serving the model take less time)
- Full-parameter shines if:
    - You want to make sure that your fine-tuned model has the maximum quality
    - You want to serve only one fine-tuned version of the model
```